processing/src/main/java/io/druid/query/groupby/epinephelinae/ConcurrentGrouper.java

```diff
@@ -292,19 +325,413 @@ private boolean isParallelSortAvailable()
                    .collect(Collectors.toList());
   }
 
+  /**
+   * Merge dictionaries of {@link Grouper.KeySerde}s of {@link Grouper}s.  The result dictionary contains unique string
+   * string keys.
+   *
+   * @return merged dictionary if its size does not exceed max dictionary size.  Otherwise null.
+   */
+  @Nullable
+  private List<String> mergeDictionary()
+  {
+    final long maxDictionarySize = combineKeySerdeFactory.getMaxDictionarySize();
+    final Set<String> mergedDictionary = new HashSet<>();
+    long totalDictionarySize = 0L;
+
+    for (SpillingGrouper<KeyType> grouper : groupers) {
+      final List<String> dictionary = grouper.getDictionary();
+      totalDictionarySize += dictionary.stream().mapToLong(RowBasedGrouperHelper::estimateStringKeySize).sum();
+      if (totalDictionarySize > maxDictionarySize) {
+        return null;
+      }
+      mergedDictionary.addAll(dictionary);
+    }
+
+    return ImmutableList.copyOf(mergedDictionary);
+  }
+
+  /**
+   * Build a combining tree for the input iterators which combine input entries asynchronously.  This method is called
+   * when data are spilled and thus streaming combine is preferred to avoid too many disk accesses.
+   *
+   * @param sortedIterators sorted iterators
+   *
+   * @return an iterator of the root grouper of the combining tree
+   */
+  private Iterator<Entry<KeyType>> parallelCombine(
+      List<Iterator<Entry<KeyType>>> sortedIterators,
+      List<String> mergedDictionary
+  )
+  {
+    // CombineBuffer is initialized when this method is called
+    final ResourceHolder<ByteBuffer> combineBufferHolder = combineBufferSupplier.get();
+    final ByteBuffer combineBuffer = combineBufferHolder.get();
+    final AggregatorFactory[] combiningFactories = new AggregatorFactory[aggregatorFactories.length];
+    for (int i = 0; i < aggregatorFactories.length; i++) {
+      combiningFactories[i] = aggregatorFactories[i].getCombiningFactory();
+    }
+    final int minimumRequiredBufferCapacity = StreamingMergeSortedGrouper.requiredBufferCapacity(
+        combineKeySerdeFactory.factorizeWithDictionary(mergedDictionary),
+        combiningFactories
+    );
+    final Pair<Integer, Integer> degreeAndBufferNum = findCombineDegreeAndNumBuffers(
+        combineBuffer,
+        minimumRequiredBufferCapacity,
+        concurrencyHint,
+        sortedIterators.size()
+    );
+
+    if (degreeAndBufferNum == null) {
```

Comment:
If there is no other sensible handling of `findCombineDegreeAndNumBuffers()` than throwing exception if it returns null, suggested to throw exception from that method.

Reply:
Changed to throw an exception from findCombineDegreeAndNumBuffers().